### DISCUSSION TITLE
Broker: Configuration of Brokers

### DISCUSSION THREAD

**Discussion by AbuHuraira537** [hassanhabib/The-Standard#257] [Q&A]
Disclaimer : I am here with very tiny experience and little knowledge
Hi @hassanhabib!
After reading the standards I  find this book very easy as it is written specifically for me I have learned a lot. After reading this book I want to implement new piece of code using these Standards but I am stuck on first phase of development of brokers due to lack of my understanding, I got some questions.
I want to ask about the standard way of fetching/retrieving configuration for brokers from database or from cloud. what standard way we should follow.
As we can not add brokers in brokers as dependencies and obliviously not any service as well (what my understanding about brokers).
I have multitenant application and every tenant communicate with external system  and their configuration lies in database how can I retrieve those configuration for Api Broker.

Other application has secretes on cloud(i.e Azure directive). How can I get those secretes from cloud in standard way. 
Both questions are similar but I want to know about standard way for Broker configurations which have dependencies.

My Initial plan to solve this is to use Standards in this way
for getting configuration from database
-> Create StorageBroker with configuration entity
-> implement StorageFoundationService 
-> Inject this service in Broker
Is it fine or I need to do in other way suggested by The Standard(community)

**Comment by Arnab-Developer** [comment]
ASP.NET has [configuration](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0) model in-built. Can we not use that?

**Reply by Arnab-Developer** [reply]
I am trying to provide an example with ASP.NET. Here I am considering `tenant` as a configuration for the broker. I can add the value of the `tenant` configuration in the `appsettings.json` file or any other supported sources by ASP.NET configuration model.

Program.cs
```csharp
WebApplicationBuilder builder = WebApplication.CreateBuilder(args);

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();
builder.Services.AddTransient<IHelloWorldService, HelloWorldService>();
builder.Services.AddTransient<ITenantBroker, TenantBroker>();

WebApplication app = builder.Build();

if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.MapGet("/weatherforecast", (IHelloWorldService service) =>
{
    return service.GetMessage();
})
.WithName("GetWeatherForecast")
.WithOpenApi();

app.Run();
```
Service layer:
```csharp
public interface IHelloWorldService
{
    public string GetMessage();
}

public class HelloWorldService : IHelloWorldService
{
    private readonly ITenantBroker broker;

    public HelloWorldService(ITenantBroker broker)
    {
        this.broker = broker
    }

    public string GetMessage()
    {
        string tenant = this.broker.GetTenant();
        return $"Hello from {tenant}";
    }
}
```
Broker layer:
```csharp
public interface ITenantBroker
{
    public string GetTenant();
}

public class TenantBroker : ITenantBroker
{
    private readonly IConfiguration configuration;

    public TenantBroker(IConfiguration configuration)
    {
        this.configuration = configuration
    }

    public string GetTenant()
    {
        string tenant = this.configuration.GetValue<string>("Tenant");
        return tenant;
    }
}
```

**Reply by Arnab-Developer** [reply]
I got a similar approach in the [documentation](https://github.com/hassanhabib/The-Standard/wiki/Brokers#4-own-their-configurations).

**Reply by LBoullosa** [reply]
You could inject IConfiguration in the broker, but you can´t have an if inside the broker.

**Reply by Arnab-Developer** [reply]
I have updated example code as per as the suggestion from @LBoullosa.

**Reply by LBoullosa** [reply]
the return as some logic, how do you test drive it? this message composition should be in the service. The Broker could be a ConfigurationBroker and just retrieves an string configuration.

Makes sense?

**Reply by Arnab-Developer** [reply]
@LBoullosa I have updated the example code.

**Reply by LBoullosa** [reply]
Ok, better, The only thing I notice you use the new way to define the class with constructor. You could use it, but it´s not Standard compliant at the moment. I remember @hassanhabib explaining something about it in some video but I don´t know which one.

**Reply by Arnab-Developer** [reply]
@LBoullosa the example code updated with normal constructor.
